src/libsync/clientsideencryption.h

```diff
@@ -200,21 +214,27 @@ class OWNCLOUDSYNC_EXPORT FolderMetadata {
 
     [[nodiscard]] bool moveFileDropToFiles();
 
+    bool addUser(const QString &userId, const QSslCertificate certificate);
+    bool removeUser(const QString &userId);
+
 private:
     /* Use std::string and std::vector internally on this class
      * to ease the port to Nlohmann Json API
      */
     void setupEmptyMetadata();
     void setupExistingMetadata(const QByteArray& metadata);
+    void updateUsers(const QByteArray &metadataKey);
 
     [[nodiscard]] QByteArray encryptData(const QByteArray &data) const;
+    [[nodiscard]] QByteArray encryptData(const QByteArray &data, const QSslKey key) const;
     [[nodiscard]] QByteArray decryptData(const QByteArray &data) const;
 
     [[nodiscard]] QByteArray encryptJsonObject(const QByteArray& obj, const QByteArray pass) const;
     [[nodiscard]] QByteArray decryptJsonObject(const QByteArray& encryptedJsonBlob, const QByteArray& pass) const;
 
     QVector<EncryptedFile> _files;
-    QMap<int, QByteArray> _metadataKeys;
+    QByteArray _metadataKey;
+    QHash<QString, FolderUser> _folderUsers;
```

Comment:
@mgallien Think of a usual Nextcloud instance with some users in it. I can not predict how many users will be registered there. It can be just 2 users - you and your family member, while it can also be a company server with any number from 100 to 10K, yes I checked the documentation for QMap vs QHash